### PR TITLE
unit tests: move to pytest

### PR DESCRIPTION
**Pull request by sergiusens**
All tests have been migrated to work with pytest.
This means that testscenarios as such is not longer being used for unit
tests and using a shim that parametrizes tests for pytest.

For the cases where the unit.TestCase was to ingrained in the test, the
strategy to replace scenarios has been to go with test duplication. This
has the benefit that testscenarios removed where it is now possible to
run an individual test.

pytest specific fixtures were added to better support some test cases.
Most tests are unchanged, but would benefit from smaller fixtures as
those added by pytest for faster test setup.

runtests.sh has been adapted to use pytest with pytest-cov, but keeping
unittest as the runner for integration tests.

The spread tests were also updated, but it would be good to see them not
run in spread anymore.

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: tests/unit/project/test_schema.py [hunk 1]
```diff
@@ -319,246 +305,250 @@ def test_valid_types(self):
...
+        "한글",
+        "ру́сский язы́к",
+        "ໄຂ່​ອີ​ສ​ເຕີ້",
+        ":a",
+        "a:",
+        "a:a",
+        "_a",
+        "a_",
+        "a_a",
+        "Hi",
```
Comment (cjp256) [inline tests/unit/project/test_schema.py:515]: Hi is dupe

File: tests/unit/project_loader/extensions/test_utils.py [hunk 2]
```diff
@@ -235,11 +233,56 @@ def test_extension_merge(self):
...
 
+    def test_extension_merge_plug_plug2(self):
+        self.assert_extensions("[plug, plug2]")
+
+    def test_extension_merge_plug2_plug(self):
+        self.assert_extensions("[plug2, plug]")
+
 
 class ExtensionMergeTest(ExtensionTestBase):
-    scenarios = [
```
Comment (cjp256) [inline tests/unit/project_loader/extensions/test_utils.py:240]: parameterization?

File: tests/unit/project_loader/grammar/test_on_statement.py [hunk 3]
```diff
@@ -226,45 +224,41 @@ class OnStatementInvalidGrammarTestCase(GrammarBaseTestCase):
         (
             "not even close",
```
Comment (cjp256) [inline tests/unit/project_loader/grammar/test_on_statement.py:225]: lol

File: tests/unit/project_loader/test_build_packages.py [hunk 4]
```diff
@@ -14,66 +14,58 @@
...
-            """\
-            name: test
-            base: core18
-            version: "1.0"
-            summary: test
-            description: test
-            confinement: strict
-            grade: stable
+def get_project_config(snapcraft_yaml_content):
+    snapcraft_yaml_path = pathlib.Path("Snapcraft.yaml")
```
Comment (cjp256) [inline tests/unit/project_loader/test_build_packages.py:27]: s/Snapcraft.yaml/snapcraft.yaml

File: tests/unit/build_providers/multipass/test_multipass_command.py [hunk 5]
```diff
@@ -54,36 +56,34 @@ def test_ensure_multipass_on_command_found(self):
...
 
-            try:
-                if self.snap_which:
-                    return ["snap"]
-            except AttributeError:
-                return []
+            return None
 
-        self.which_mock.side_effect = which_effect
+        monkeypatch.setattr(shutil, "which", which_effect)
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/build_providers/multipass/test_multipass_command.py:82]: We should avoid unittest.mock at all cost and prefer dependency injection where possible, this should be red flag for anything new we design.

File: tests/unit/commands/test_collaborators.py [hunk 6]
```diff
@@ -45,10 +45,7 @@ def setUp(self):
 
 class CollaboratorsCommandTestCase(CollaborateBaseTestCase):
 
-    scenarios = [
-        ("edit-collaborators", {"command_name": "edit-collaborators"}),
-        ("collaborators alias", {"command_name": "collaborators"}),
-    ]
+    command_name = "edit-collaborators"
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/commands/test_collaborators.py:48]: agreed

File: tests/unit/commands/test_init.py [hunk 7]
```diff
@@ -63,46 +63,20 @@ def test_init_must_write_snapcraft_yaml(self):
...
-        (
-            "snap/snapcraft.yaml",
-            {
-                "yaml_path": os.path.join("snap", "snapcraft.yaml"),
-                "message": "snap/snapcraft.yaml already exists!\n",
-            },
-        ),
-        # FIXME: Both snapcraft.yaml and .snapcraft.yaml are deprecated.
-        (
-            "snapcraft.yaml",
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/commands/test_init.py:77]: Yeah, we have this tested in a different module

File: tests/unit/conftest.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    for scenario in metafunc.cls.scenarios:
+        idlist.append(scenario[0])
+        items = scenario[1].items()
+        argnames = [x[0] for x in items]
+        argvalues.append([x[1] for x in items])
+    metafunc.parametrize(argnames, argvalues, ids=idlist, scope="class")
+
+
+@pytest.fixture
+def tmp_work_path(tmp_path):
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/conftest.py:40]: pytest builtin fixtures

[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/conftest.py:40]: I should add a docstring to this

File: tests/unit/plugins/v1/test_catkin.py [hunk 9]
```diff
@@ -1153,102 +1142,354 @@ def test_pull_pip_dependencies(self, generate_setup_mock):
...
+            f"{plugin.builddir}/src",
+            "--install-space",
+            f"{plugin.installdir}/opt/ros/melodic",
+            "-j{}".format(1 if disable_parallel else 2),
+            "--cmake-args",
+            f"-DCMAKE_BUILD_TYPE={build_type}",
+        ]
+        if catkin_cmake_args:
+            expected_command += catkin_cmake_args
+        assert run_mock.mock_calls[0][1][0] == expected_command
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/plugins/v1/test_catkin.py:1492]: agreed to use assert_has_calls

File: tests/unit/plugins/v1/test_nodejs.py [hunk 10]
```diff
@@ -620,15 +723,19 @@ def test_unsupported_arch_raises_exception(self):
...
 
 class NodePluginMissingFilesTest(NodePluginBaseTest):
-
-    scenarios = [
-        ("npm", dict(package_manager="npm")),
-        ("yarn", dict(package_manager="yarn")),
-    ]
-
-    def test_missing_package_json(self):
+    def run_test(self):
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/plugins/v1/test_nodejs.py:726]: no promises, but I will look into it

File: tests/unit/plugins/v1/test_rust.py [hunk 11]
```diff
@@ -135,162 +149,77 @@ class RustPluginCrossCompileTest(RustPluginBaseTest):
...
+        target,
+    ):
+        monkeypatch.setattr(snapcraft.project.Project, "is_cross_compiling", True)
+        project = snapcraft.project.Project(target_deb_arch=deb_arch)
+        project._snap_meta = meta.snap.Snap(name="test-snap", base="core18")
+
+        plugin = rust.RustPlugin("test-part", options, project)
+        plugin.pull()
+
+        assert len(mock_run.mock_calls) == 4
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/plugins/v1/test_rust.py:171]: moving to assert_has_calls

File: tests/unit/project/test_schema.py [hunk 12]
```diff
@@ -192,50 +191,44 @@ def test_missing_required_property_and_missing_adopt_info(self):
         self.assertThat(raised.message, Equals(expected_message), message=self.data)
 
 
-class DaemonDependencyTest(ValidationBaseTest):
+def test_icon_missing_is_valid_yaml(monkeypatch, data):
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/project/test_schema.py:194]: I will remove this test

File: tests/unit/project_loader/extensions/test_utils.py (hunk 2 above)
Comment (sergiusens) [reply tests/unit/project_loader/extensions/test_utils.py:240]: maybe later, this module has too many fixtures

**Pull request merged**